pkg/chunk/repair.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package chunk
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"strings"
+
+	"github.com/ethersphere/bee/pkg/api"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/pss"
+	"github.com/ethersphere/bee/pkg/pushsync"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/trojan"
+)
+
+const (
+	// RecoveryTopicText is the string used to construct the recovery topic
+	RecoveryTopicText = "RECOVERY"
+)
+
+var (
+	// RecoveryTopic is the topic used for repairing globally pinned chunks
+	RecoveryTopic = trojan.NewTopic(RecoveryTopicText)
+
+	// ErrTargetPrefix is returned when target prefix decoding fails
+	ErrTargetPrefix = errors.New("error decoding prefix string")
+)
+
+// RecoveryHook defines code to be executed upon failing to retrieve pinned chunks
+type RecoveryHook func(ctx context.Context, chunkAddress swarm.Address) error
+
+// sender is the function call for sending trojan chunks
+type sender func(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*pss.Monitor, error)
+
+// NewRecoveryHook returns a new RecoveryHook with the sender function defined
+func NewRecoveryHook(send sender) RecoveryHook {
+	return func(ctx context.Context, chunkAddress swarm.Address) error {
+		targets, err := getPinners(ctx)
+		if err != nil {
+			return err
+		}
+		payload := chunkAddress
+
+		if _, err := send(ctx, targets, RecoveryTopic, payload.Bytes()); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+// NewRepairHandler creates a repair function to re-upload globally pinned chunks to the network with the given store
+func NewRepairHandler(s storage.Storer, logger logging.Logger, pushSyncer pushsync.PushSyncer) pss.Handler {
+	return func(m trojan.Message) {
+		chAddr := m.Payload
+		ctx := context.Background()
+
+		ch, err := s.Get(ctx, storage.ModeGetRequest, swarm.NewAddress(chAddr))
+		if err != nil {
+			logger.Debugf("chunk repair: error while getting chunk for repairing: %v", err)
+			return
+		}
+
+		// push the chunk using push sync so that it reaches it destination in network
+		_, err = pushSyncer.PushChunkToClosest(ctx, ch)
+		if err != nil {
+			logger.Debugf("chunk repair: error while sending chunk or receiving receipt: %v", err)
+			return
+		}
+	}
+}
```

Comment:
what would happen if after calling `s.Get` the chunk is garbage collected due to not being locally pinned?

Reply:
well the repair will print the error and return. I think this is OK. I would even go as far as putting the debug line in the error handling block to trace

Reply:
that's fair, but i don't think we should be repairing chunks which are not locally pinned.

to me, it seems kind of counter intuitive to mark chunks as globally pinned which we are not pinning locally, but we can certainly discuss this further.

Reply:
> but i don't think we should be repairing chunks which are not locally pinned.

Why? Is this a hard requirement? as long as the pinner is able to access the chunk.. it should repair it.

Reply:
I think this is an edge-condition. Normally, you would expect a chunk to be locally pinned, but if a chunk is not locally pinned and the "pinner" can still repair, then why not.

Reply:
The pinner might use ways outside of Swarm to make sure a chunk is persisted. Hence, a chunk might not be locally pinned from Swarm-perspective, but it is still persisted by the node. I think it's fine for a chunk not to be locally pinned, but still repair-able.

Reply:
why would we advertise content as globally pinned which is not being pinned locally?

if a node goes through the trouble of enabling itself as a global pinner and having its address advertised, i don't think it's too much to ask to also pin the content it's supposed to be able to restore.

let's also look at the flip side: if we don't force local pinning here, we are potentially having a ton of wasted repair requests, traffic, etc.

finally, with regards to outside mechanisms of chunk persistence, i suggest we cross that bridge when we get to it.

Reply:
As i said in the `mattermost`, we should recommend the pinner's to pin it for providing proper service. But we don't have to check if it is pinned. We just have to see if it is accessible (which is the same). If we can get the chunk, we repair it, otherwise not.

Reply:
i understand yours and @acud's points, but i still think this is conceptually unsound.

it seems i'm the only one who is insisting on this though, so i will gladly resolve this thread and not block on this any longer.